### DISCUSSION TITLE
Is the memory usage lower for NetCDFOutputWriter compared to JLD2OutputWriter on GPU?

### DISCUSSION THREAD

**Discussion by liuchihl** [CliMA/Oceananigans.jl#3533] [Computational science!]
My simulation, which is (150x300x100) in size, was running on a GPU V100. After 532 iterations, which took about 1 hour of wall time, it terminated due to an out-of-memory (OOM) error when I was using the JLD2OutputWriter to write the output. However, when I switched to using the NetCDFOutputWriter for the exact same simulation and diagnostics, the simulation completed in about 10 minutes of wall time (2219 iterations) without any memory issues. I'm unsure what caused this significant difference in memory usage between the two output writers. Does this issue specifically occur for large (or not-small) simulations on GPUs? I'm pleased with how much more efficient the NetCDF output writer is, but I'd like to understand the reason behind this difference. Any insight would be much appreciated! Thank you!

(I am using Oceananigans v.90.9 and Julia 1.9.3)

**Comment by glwagner** [comment]
Do you mind sharing your script? Both the outputs and the model configuration can affect how the output writers behave.

There are not any major differences in memory consumption between the two output writers that I know of. There may be some small differences because they are initialized differently (with `JLD2`, we serialize some model properties for convenience). So, your experience could be indicative of a bug. I also don't think there is any reason why the simulations would involve different numbers of iterations (if the `schedule` is the same, the number of iterations should be similar modulo any randomness in the initial condition).

**Comment by liuchihl** [comment]
Sure, no problem, here is the [script](https://github.com/liuchihl/internal-tide-mixing/blob/45bc947888d77db9926d330ef1e686431ce91413/tilted_internal_tide_experiment.jl).

Yes, the `schedule` is identical for both writers. The difference lies in the fact that with JLD2OutputWriter, the model stopped at iteration 532, whereas with the NetCDF writer, it completed 2219 iterations, which is the number of iterations to finish this example.

**Reply by glwagner** [reply]
Ah ok, I understand what you're saying.

Nothing jumps out at me after reading your script! It all looks ok. A puzzle!

Is the output as expected / have the right size? Perhaps another good-to-know info would be: after 532 iterations, what is the size of the `.jld2` files versus the size of the `.nc` files? 

I can only guess that the serialization used by JLD2OutputWriter is misbehaving...

@simone-silvestri might have something useful to add. I think he has used `JLD2OutputWriter` for some absolutely mind bogglingly large simulations with no issues as far as I know...

Hmm, I guess another useful piece of information is what filesystem / machine you are using.

Here's another test: configure all the output writers to save every iteration, and then set up the simulation to run for 1 or 2 iterations saving just a few snapshots. Run the simulation both on your laptop and on the cluster (assuming you are using one). Look to see if the size of the output is identical on your laptop CPU as on the cluster to see if the cluster filesystem is playing a role.

**Reply by liuchihl** [reply]
> Hmm, I guess another useful piece of information is what filesystem / machine you are using.

There are multiple DFS systems with an aggregate throughput of more than 30 GB/s. The machine I am using is [HPC3](https://rcic.uci.edu/hpc3/hpc3.html). 

Thanks for the suggestions, I'll do some tests, it would be interesting to know the answer!

**Reply by liuchihl** [reply]
Ok, I have done two experiments on a cluster using GPU, running only 3 iterations of the same simulation as above.

(1) writing output via JLD2outputwriter: 
- simulation initialization complete (36.679 minutes).   (which is way much longer than I expected)
- output file size=723.9 MiB
- took ~40 mins of wallclock to complete the job

(2) writing output via NetCDFoutputwriter
- simulation initialization complete (34.732 seconds)
- output file size=268.5 MiB
- 4 mins of wallclock time to complete the job

Seems like the culprit might be associated with simulation initialization when using JLD2outputwriter. Netcdfoutputwriter initializes just fine.  
This reminds me of issue #3381, where the slow initialization can be resolved by upgrading both Julia and Oceananigans.

**Reply by glwagner** [reply]
40 minutes is crazy! I think we are hitting some bad compile path. Is the 40 min only on the cluster, or do you also hit that bad path on your laptop / locally? I agree, this does remind me of #3381. The weird thing is that it's only JLD2 though, whereas I think that other issue occurred for both (it seemed like more of a general problem with outputs).

Sorry to add work, but I guess just the raw filesize is not enough, we also need to know the difference between 1 and 2 iterations to know whether JLD2OutputWriter is including a lot more data per iteration. The absolute data will definitely be different because we save a bit more in JLD2OutputWriter (we serialize the entire grid).

In particular I don't think the `bottom_height` for the grid is saved in `NetCDFOutputWriter`. Is that right?

**Reply by liuchihl** [reply]
I haven't tried it on my laptop yet, but I do have access to two different clusters. Both of them have long initialization time (~30 mins) when using Julia version below 1.10 with `JLD2outputwriter`. 

No worries, and thanks for your kind help! Here are the output files for both outputwriters: $\Delta$ size seems to be 4 times longer when using `JLD2OutputWriter` than `NetCDFOutputWriter`. 
`JLD2Outputwriter`: 
<img width="967" alt="Screenshot 2024-04-03 at 11 38 09 AM" src="https://github.com/CliMA/Oceananigans.jl/assets/68127124/8dcf276a-cd28-4b2a-b7b2-a7ed3ed672c5">

`NetCDFoutputwriter`: 
<img width="891" alt="Screenshot 2024-04-03 at 11 37 21 AM" src="https://github.com/CliMA/Oceananigans.jl/assets/68127124/fef9f4ff-c5ce-46e6-9f14-e7c2aff3ed0a">

I think you are right, I couldn't find the `bottom_height` in my output file when using `NetCDFOutputWriter`. 
Is it possible to include `bottom_height` in the netcdfoutput?

**Reply by glwagner** [reply]
> Is it possible to include bottom_height in the netcdfoutput?

Yes for sure. I'm not sure the best way to do this... we need an interface for saving the properties of `ImmersedBoundaryGrid`, because different immersed boundary methods have different properties (for example, sometimes we have `bottom_height` but other times we just have a 3D `mask` for a 3D immersed boundary).

Probably we should open an issue to discuss how to save properties of `ImmersedBoundaryGrid` with `NetCDFOutputWriter`. I am happy to do that but since you brought it up, you should feel free to open it yourself!

I just noticed that your `JLD2OutputWriter` is saving `all_diags` for fields:

https://github.com/liuchihl/internal-tide-mixing/blob/45bc947888d77db9926d330ef1e686431ce91413/tilted_internal_tide_experiment.jl#L189

However, `NetCDFOutputWriter` is only saving `custom_diags`:

https://github.com/liuchihl/internal-tide-mixing/blob/45bc947888d77db9926d330ef1e686431ce91413/tilted_internal_tide_experiment.jl#L212

Could this be the cause of the discrepancy? 

Interesting that "all diags" takes way longer to compile than just "custom diags". It doesn't really make sense.

**Reply by liuchihl** [reply]
Sure I can open an issue.

> Could this be the cause of the discrepancy?

Ah, you're absolutely right, I should have saved the same diagnostics. That is indeed the discrepancy. After using `all_diags`, the initialization time for `NetCDFOutputWriter` increased to pretty much the same as `JLD2OutputWriter` (which is ~36 minutes). 
This long initialization issue is consistent with what people have found in #3381.  

> Interesting that "all diags" takes way longer to compile than just "custom diags". It doesn't really make sense.

This is very surprising, but when upgrading to Julia 1.10, this long-initialization issue no longer persists when using both writers.

**Reply by glwagner** [reply]
Ok, I should have read more closely the first time!! Glad we caught the issue though.

To summarize:

- Memory usage is indeed fairly similar between NetCDF and JLD2 output writers
- It'd be nice to save aspects of the immersed boundary grid when using NetCDF
- There's a long initialization issue on 1.9 but it's solved by 1.10, as far as we can tell...

**Reply by liuchihl** [reply]
Thanks for spotting that!!

I think I actually found the main issue of my case: 
I noticed the way I define my diagnostics is to merge different types of data arrays. For example, `model.velocities` and `model.tracers` are different from `custom_diags = (B=B, uhat=û, what=ŵ)`:
https://github.com/liuchihl/internal-tide-mixing/blob/45bc947888d77db9926d330ef1e686431ce91413/tilted_internal_tide_experiment.jl#L177

The long initialization issue has been resolved, and now only takes a few seconds once I separate the two output writing processes. My guess is that the output writer operates most efficiently when the diagnostics are consistent in types, though merging them into just one file would be ideal.

**Reply by glwagner** [reply]
That does seem like a culprit --- mixing `Field` with `AbstractOperation`.

That said, users _should_ be able to mix these two. It's inconvenient to have to use two different output writers! I bet we can fix things by using dictionaries rather than `NamedTuple` for outputs --- or some silly thing like that.

Do you still have long initialization time on julia 1.10?

**Reply by liuchihl** [reply]
No, the good news is that there is no long initialization time on Julia 1.10 even when mixing different types of diagnostics.

**Comment by simone-silvestri** [comment]
Could you also share the error? The GPU should not be a problem as the output is transferred to the CPU before writing. If it's an OOM error it means we are consuming too much RAM, but, as far as I know, there should be no heavy memory allocation in the output writers so it might mean that there is a bug somewhere. 
Knowing the line where the failure happens might indicate where the memory allocation happens

**Reply by liuchihl** [reply]
Yes, of course. Below are the final lines of the output message along with the error when using JLD2outputwriter: 
```
[ Info: Fetching JLD2 output (:u, :v, :w, :b, :B, :uhat, :what)...
[ Info: Fetching time: 116.340 ms
[ Info: Writing JLD2 output (:u, :v, :w, :b, :B, :uhat, :what) to ./output/internal_tide_3days-theta=0.0036_realtopo3D_Nx150_fields.jld2...
[ Info: Writing done: time=1.047 seconds, size=10.1 GiB, Δsize=241.0 MiB
[ Info: Fetching JLD2 output (:B, :uhat, :what)...
[ Info: Fetching time: 502.184 μs
[ Info: Writing JLD2 output (:B, :uhat, :what) to ./output/internal_tide_3days-theta=0.0036_realtopo3D_Nx150_slices.jld2...
[ Info: Writing done: time=49.503 ms, size=15.6 MiB, Δsize=353.1 KiB
/export/spool/slurm/slurmd.spool/job30694661/slurm_script: line 21: 1106691 Killed                  julia --project $rundir > $rundir_out.txt
slurmstepd: error: Detected 1 oom_kill event in StepId=30694661.batch. Some of the step tasks have been OOM Killed.
```

> Knowing the line where the failure happens might indicate where the memory allocation happens

I agree, but it's not clear to me where the failure might have occurred.
